### ISSUE TITLE
cocos2dx v3.11 [lua] [Android] audio.playMusic() cannot find mp3 file

### ISSUE THREAD

**Issue by jie-meng** [cocos2d/cocos2d-x#15782]
cocos2dx version: v3.11
language: lua
platform: android

mp3 was put under 'res' dir.

audio.playMusic() can not make any music, adb log shows java.io cannot find file exception.

audio.playSound() works well.

On desktop

**Comment by jie-meng** [comment]
I also test it with cocos2dx version v3.11.1, bug still there.

**Comment by jie-meng** [comment]
```
06-04 03:03:05.105 31754 31773 E Cocos2dxMusic: error: boss.mp3
06-04 03:03:05.105 31754 31773 E Cocos2dxMusic: java.io.FileNotFoundException: boss.mp3
06-04 03:03:05.105 31754 31773 E Cocos2dxMusic:     at android.content.res.AssetManager.openAssetFd(Native Method)
06-04 03:03:05.105 31754 31773 E Cocos2dxMusic:     at android.content.res.AssetManager.openFd(AssetManager.java:333)
06-04 03:03:05.105 31754 31773 E Cocos2dxMusic:     at org.cocos2dx.lib.Cocos2dxMusic.createMediaplayer(Cocos2dxMusic.java:253)
06-04 03:03:05.105 31754 31773 E Cocos2dxMusic:     at org.cocos2dx.lib.Cocos2dxMusic.playBackgroundMusic(Cocos2dxMusic.java:96)
06-04 03:03:05.105 31754 31773 E Cocos2dxMusic:     at org.cocos2dx.lib.Cocos2dxHelper.playBackgroundMusic(Cocos2dxHelper.java:235)
06-04 03:03:05.105 31754 31773 E Cocos2dxMusic:     at org.cocos2dx.lib.Cocos2dxRenderer.nativeInit(Native Method)
06-04 03:03:05.105 31754 31773 E Cocos2dxMusic:     at org.cocos2dx.lib.Cocos2dxRenderer.onSurfaceCreated(Cocos2dxRenderer.java:72)
06-04 03:03:05.105 31754 31773 E Cocos2dxMusic:     at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1503)
06-04 03:03:05.105 31754 31773 E Cocos2dxMusic:     at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1240)
06-04 03:03:05.105 31754 31773 E Cocos2dxMusic: playBackgroundMusic: background media player is null
06-04 03:03:05.105 31754 31773 D cocos2d-x debug info: create rendererRecreatedListener for GLProgramState
```

**Comment by jie-meng** [comment]
I checked the source of cocos2dx 3.9 (which is ok) and 3.11.1, found createMediaplayer(final String path) method are different.

3.9

```
private MediaPlayer createMediaplayer(final String path) {
        MediaPlayer mediaPlayer = new MediaPlayer();

        try {
            if (CocosPlayClient.isEnabled() && !CocosPlayClient.isDemo()) {
                CocosPlayClient.updateAssets(path);
            }
            CocosPlayClient.notifyFileLoaded(path);
            if (path.startsWith("/")) {
                final FileInputStream fis = new FileInputStream(path);
                mediaPlayer.setDataSource(fis.getFD());
                fis.close();
            } else {
                final AssetFileDescriptor assetFileDescritor = this.mContext.getAssets().openFd(path);
                mediaPlayer.setDataSource(assetFileDescritor.getFileDescriptor(), assetFileDescritor.getStartOffset(), assetFileDescritor.getLength());
            }

            mediaPlayer.prepare();

            mediaPlayer.setVolume(this.mLeftVolume, this.mRightVolume);
        } catch (final Exception e) {
            mediaPlayer = null;
            Log.e(Cocos2dxMusic.TAG, "error: " + e.getMessage(), e);
        }

        return mediaPlayer;
    }
```

3.11.1

```
private MediaPlayer createMediaplayer(final String path) {
        MediaPlayer mediaPlayer = new MediaPlayer();

        try {
            if (path.startsWith("/")) {
                final FileInputStream fis = new FileInputStream(path);
                mediaPlayer.setDataSource(fis.getFD());
                fis.close();
            } else {
                final AssetFileDescriptor assetFileDescritor = this.mContext.getAssets().openFd(path);
                mediaPlayer.setDataSource(assetFileDescritor.getFileDescriptor(), assetFileDescritor.getStartOffset(), assetFileDescritor.getLength());
            }

            mediaPlayer.prepare();

            mediaPlayer.setVolume(this.mLeftVolume, this.mRightVolume);
        } catch (final Exception e) {
            mediaPlayer = null;
            Log.e(Cocos2dxMusic.TAG, "error: " + e.getMessage(), e);
        }

        return mediaPlayer;
    }

```

These lines removed from v3.11

```
if (CocosPlayClient.isEnabled() && !CocosPlayClient.isDemo()) {
                CocosPlayClient.updateAssets(path);
            }
            CocosPlayClient.notifyFileLoaded(path);
```

And 

```
this.mContext.getAssets().openFd(path)
```

Does this line indicate that I should use resource(image/music) with a relative path from the directory "assets" instead of "res"?

I change 

```
audio.playMusic("boss.mp3", true)
```

to

```
audio.playMusic("res/boss.mp3", true)
```

then it works.

So why does the new-create-project use 

```
display.newSprite("HelloWorld.png")
```

Which gives me hint of using resource relative to directory "assets/res" ??

**Comment by minggo** [comment]
No, resources is relative to `Resources` folder. On Android, all resources are copied to `assets`.

**Comment by jie-meng** [comment]
@minggo 

I put "boss.mp3" under the path 'res'.

Then I call audio.PlayMusic("boss.mp3", true), There is no music on cocos2dx V3.11, but on V3.9 it's OK, please check it. (Android platform)

**Comment by minggo** [comment]
> There is no music on cocos2dx V3.11, but on V3.9 it's OK, please check it. (Android platform)

I think it is a bug on v3.9 because it is different from other platforms.

**Comment by jie-meng** [comment]
@minggo 
I don't think it is bug of v3.9, because I update my project on 3.5, 3.6, 3.8, 3.9, they are all ok, until I update it to 3.11, BGM cannot play.

So I tried to check the c++ code in framework and I found in 3.11

```
if (CocosPlayClient.isEnabled() && !CocosPlayClient.isDemo()) {
                CocosPlayClient.updateAssets(path);
            }
```

was removed from method 

```
private MediaPlayer createMediaplayer(final String path)
```

Then I change 'audio.playMusic("boss.mp3", true)' to 'audio.playMusic("res/boss.mp3", true)', BGM works.

**Comment by minggo** [comment]
@joshua-meng In which file is the codes removed?

**Comment by jie-meng** [comment]
@minggo 

frameworks/cocos2d-x/cocos/platform/android/java/src/org/cocos2dx/lib/Cocos2dxMusic.java

In function

```
private MediaPlayer createMediaplayer(final String path)
```

**Comment by minggo** [comment]
Yep, i found it. The codes are something related with cocosplay. I think it is good to remove them. And pass the path will full path relative to `Resources` is more reasonable.

And i think the reason that `audio.playMusic()` can work is because [main.lua](https://github.com/cocos2d/cocos2d-x/blob/v3/templates/lua-template-default/src/main.lua) adds `res` into search path. And the codes keep the same as before. 

Did you modify `main.lua` that make it not work?

**Comment by jie-meng** [comment]
No I did not modify it at all.

```
cc.FileUtils:getInstance():setPopupNotify(false)
cc.FileUtils:getInstance():addSearchPath("src/")
cc.FileUtils:getInstance():addSearchPath("res/")

require "config"
require "cocos.init"

local function main()
    require("app.MyApp"):create():run()
end

local status, msg = xpcall(main, __G__TRACKBACK__)
if not status then
    print(msg)
end
```

You can reproduce this bug easily by
- Create a new lua project "cocos new Test -p com.example.test -l lua"
- Put a "test.mp3" file under "res"
- Call `audio.playMusic("test.mp3")` On MainScene onCreate function

There's no music playing on v3.11.1

**Comment by minggo** [comment]
@joshua-meng I think it is fixed in [this commit](https://github.com/cocos2d/cocos2d-x/pull/15512).

**Comment by jie-meng** [comment]
@minggo I checked the code in cocos/audio/android/jni/cddandroidAndroidJavaEngine.cpp and found it still not merged into v3.11.1 yet, thanks.
